File: API/2010_ColorMaRsi_Trigger/CS/ColorMaRsiTriggerStrategy.cs
using System;
using System.Collections.Generic;

using StockSharp.Algo.Indicators;
using StockSharp.Algo.Strategies;
using StockSharp.BusinessEntities;
using StockSharp.Messages;

namespace StockSharp.Samples.Strategies;

/// <summary>
/// ColorMaRsi Trigger strategy.
/// Combines fast and slow EMA and RSI to generate trading signals.
/// </summary>
public class ColorMaRsiTriggerStrategy : Strategy
{
private readonly StrategyParam<int> _emaFastLength;
private readonly StrategyParam<int> _emaSlowLength;
private readonly StrategyParam<int> _rsiFastLength;
private readonly StrategyParam<int> _rsiSlowLength;
private readonly StrategyParam<DataType> _candleType;

private ExponentialMovingAverage _emaFast = null!;
private ExponentialMovingAverage _emaSlow = null!;
private RelativeStrengthIndex _rsiFast = null!;
private RelativeStrengthIndex _rsiSlow = null!;
private decimal _prevSignal;

/// <summary>
/// Fast EMA period.
/// </summary>
public int EmaFastLength { get => _emaFastLength.Value; set => _emaFastLength.Value = value; }

/// <summary>
/// Slow EMA period.
/// </summary>
public int EmaSlowLength { get => _emaSlowLength.Value; set => _emaSlowLength.Value = value; }

/// <summary>
/// Fast RSI period.
/// </summary>
public int RsiFastLength { get => _rsiFastLength.Value; set => _rsiFastLength.Value = value; }

/// <summary>
/// Slow RSI period.
/// </summary>
public int RsiSlowLength { get => _rsiSlowLength.Value; set => _rsiSlowLength.Value = value; }

/// <summary>
/// Candle type used for calculations.
/// </summary>
public DataType CandleType { get => _candleType.Value; set => _candleType.Value = value; }

/// <summary>
/// Initializes strategy parameters.
/// </summary>
public ColorMaRsiTriggerStrategy()
{
_emaFastLength = Param(nameof(EmaFastLength), 5)
.SetGreaterThanZero()
.SetDisplay("Fast EMA", "Fast EMA period", "General")
.SetCanOptimize(true)
.SetOptimize(3, 20, 1);

_emaSlowLength = Param(nameof(EmaSlowLength), 10)
.SetGreaterThanZero()
.SetDisplay("Slow EMA", "Slow EMA period", "General")
.SetCanOptimize(true)
.SetOptimize(5, 40, 1);

_rsiFastLength = Param(nameof(RsiFastLength), 3)
.SetGreaterThanZero()
.SetDisplay("Fast RSI", "Fast RSI period", "General")
.SetCanOptimize(true)
.SetOptimize(2, 20, 1);

_rsiSlowLength = Param(nameof(RsiSlowLength), 13)
.SetGreaterThanZero()
.SetDisplay("Slow RSI", "Slow RSI period", "General")
.SetCanOptimize(true)
.SetOptimize(5, 40, 1);

_candleType = Param(nameof(CandleType), TimeSpan.FromHours(4).TimeFrame())
.SetDisplay("Candle Type", "Type of candles", "General");
}

/// <inheritdoc />
public override IEnumerable<(Security sec, DataType dt)> GetWorkingSecurities()
{
yield return (Security, CandleType);
}

/// <inheritdoc />
protected override void OnReseted()
{
base.OnReseted();
_prevSignal = 0;
}

/// <inheritdoc />
protected override void OnStarted(DateTimeOffset time)
{
base.OnStarted(time);

_emaFast = new ExponentialMovingAverage { Length = EmaFastLength };
_emaSlow = new ExponentialMovingAverage { Length = EmaSlowLength };
_rsiFast = new RelativeStrengthIndex { Length = RsiFastLength };
_rsiSlow = new RelativeStrengthIndex { Length = RsiSlowLength };

var subscription = SubscribeCandles(CandleType);
subscription
.Bind(_emaFast, _emaSlow, _rsiFast, _rsiSlow, ProcessCandle)
.Start();

var area = CreateChartArea();
if (area != null)
{
DrawCandles(area, subscription);
DrawIndicator(area, _emaFast);
DrawIndicator(area, _emaSlow);
DrawIndicator(area, _rsiFast);
DrawIndicator(area, _rsiSlow);
}
}

private void ProcessCandle(ICandleMessage candle, decimal emaFastValue, decimal emaSlowValue, decimal rsiFastValue, decimal rsiSlowValue)
{
if (candle.State != CandleStates.Finished)
return;

if (!_emaFast.IsFormed || !_emaSlow.IsFormed || !_rsiFast.IsFormed || !_rsiSlow.IsFormed)
return;

if (!IsFormedAndOnlineAndAllowTrading())
return;

var signal = 0m;
if (emaFastValue > emaSlowValue)
signal += 1;
if (emaFastValue < emaSlowValue)
signal -= 1;
if (rsiFastValue > rsiSlowValue)
signal += 1;
if (rsiFastValue < rsiSlowValue)
signal -= 1;

if (signal > 1)
signal = 1;
if (signal < -1)
signal = -1;

if (_prevSignal > 0)
{
if (Position < 0)
BuyMarket(Math.Abs(Position));

if (signal <= 0 && Position <= 0)
BuyMarket(Volume);
}
else if (_prevSignal < 0)
{
if (Position > 0)
SellMarket(Position);

if (signal >= 0 && Position >= 0)
SellMarket(Volume);
}

_prevSignal = signal;
}
}

File: API/2010_ColorMaRsi_Trigger/PY/color_ma_rsi_trigger_strategy.py
import clr

clr.AddReference("StockSharp.Messages")
clr.AddReference("StockSharp.Algo")

from System import TimeSpan
from StockSharp.Messages import DataType, CandleStates
from StockSharp.Algo.Indicators import ExponentialMovingAverage, RelativeStrengthIndex
from StockSharp.Algo.Strategies import Strategy

class color_ma_rsi_trigger_strategy(Strategy):
    """
    ColorMaRsi Trigger strategy.
    Combines fast and slow EMA crossover with RSI crossover to generate trading signals.
    """

    def __init__(self):
        super(color_ma_rsi_trigger_strategy, self).__init__()
        self._ema_fast_length = self.Param("EmaFastLength", 5) \
            .SetDisplay("Fast EMA", "Fast EMA period", "General")
        self._ema_slow_length = self.Param("EmaSlowLength", 10) \
            .SetDisplay("Slow EMA", "Slow EMA period", "General")
        self._rsi_fast_length = self.Param("RsiFastLength", 3) \
            .SetDisplay("Fast RSI", "Fast RSI period", "General")
        self._rsi_slow_length = self.Param("RsiSlowLength", 13) \
            .SetDisplay("Slow RSI", "Slow RSI period", "General")
        self._candle_type = self.Param("CandleType", DataType.TimeFrame(TimeSpan.FromHours(1))) \
            .SetDisplay("Candle Type", "Type of candles", "General")

        self._prev_signal = 0.0

    @property
    def candle_type(self):
        return self._candle_type.Value

    def OnReseted(self):
        super(color_ma_rsi_trigger_strategy, self).OnReseted()
        self._prev_signal = 0.0

    def OnStarted(self, time):
        super(color_ma_rsi_trigger_strategy, self).OnStarted(time)

        ema_fast = ExponentialMovingAverage()
        ema_fast.Length = self._ema_fast_length.Value
        ema_slow = ExponentialMovingAverage()
        ema_slow.Length = self._ema_slow_length.Value
        rsi_fast = RelativeStrengthIndex()
        rsi_fast.Length = self._rsi_fast_length.Value
        rsi_slow = RelativeStrengthIndex()
        rsi_slow.Length = self._rsi_slow_length.Value

        subscription = self.SubscribeCandles(self.candle_type)
        subscription.Bind(ema_fast, ema_slow, rsi_fast, rsi_slow, self.on_process).Start()

    def on_process(self, candle, ema_fast_val, ema_slow_val, rsi_fast_val, rsi_slow_val):
        if candle.State != CandleStates.Finished:
            return

        signal = 0.0
        if ema_fast_val > ema_slow_val:
            signal += 1
        if ema_fast_val < ema_slow_val:
            signal -= 1
        if rsi_fast_val > rsi_slow_val:
            signal += 1
        if rsi_fast_val < rsi_slow_val:
            signal -= 1

        signal = max(-1.0, min(1.0, signal))

        if self._prev_signal <= 0 and signal > 0 and self.Position <= 0:
            if self.Position < 0:
                self.BuyMarket()
            self.BuyMarket()
        elif self._prev_signal >= 0 and signal < 0 and self.Position >= 0:
            if self.Position > 0:
                self.SellMarket()
            self.SellMarket()

        self._prev_signal = signal

    def CreateClone(self):
        return color_ma_rsi_trigger_strategy()
